qiskit_experiments/calibration/experiments/rabi.py

```diff
@@ -0,0 +1,337 @@
+
+        # Sigmas may be None and fitting will not work if any sigmas are exactly 0.
+        try:
+            sigmas = y_sigmas[:, 1] / (max_y - min_y)
+            if any(sigmas == 0.0):
+                sigmas = None
+
+        except TypeError:
+            sigmas = None
+
+        xdata = np.array([datum["metadata"]["xval"] for datum in experiment_data.data()])
+
+        # Perform fit
+        best_fit = None
+
+        def fit_fun(x, amplitude, frequency, phase, offset):
+            return amplitude * np.cos(frequency * x + phase) + offset
+
+        bounds = {
+            "amplitude": amp_bounds,
+            "frequency": freq_bounds,
+            "phase": phase_bounds,
+            "offset": offset_bounds,
+        }
+
+        # Guesses have two phases to catch signals that start at 1 or 0 at zero-amplitude.
+        guesses = []
+        for freq in freq_guesses:
+            if phase_guess is None:
+                guesses.append((freq, 0))
+                guesses.append((freq, np.pi))
+            else:
+                guesses.append((freq, phase_guess))
+
+        for guess in guesses:
+            freq_guess, phase_guess = guess[0], guess[1]
+            init = {
+                "amplitude": amp_guess,
+                "frequency": freq_guess,
+                "phase": phase_guess,
+                "offset": offset_guess,
+            }
+
+            try:
+                fit_result = curve_fit(fit_fun, xdata, ydata, init, sigmas, bounds)
+
+                if not best_fit:
+                    best_fit = fit_result
+                else:
+                    if fit_result["reduced_chisq"] < best_fit["reduced_chisq"]:
+                        best_fit = fit_result
+
+            except RuntimeError:
+                pass
+
+        if best_fit is None:
+            raise QiskitError("Could not find a fit to the spectroscopy data.")
+
+        best_fit["value"] = best_fit["popt"][1]
+        best_fit["stderr"] = (fit_result["popt_err"][1],)
+        best_fit["label"] = "Spectroscopy"
+        best_fit["xdata"] = xdata
+        best_fit["ydata"] = ydata
+        best_fit["ydata_err"] = sigmas
+        best_fit["quality"] = self._fit_quality(
+            best_fit["popt"][1], best_fit["reduced_chisq"], best_fit["popt_err"][1]
+        )
+
+        if plot and plotting.HAS_MATPLOTLIB:
+            ax = plotting.plot_curve_fit(fit_fun, fit_result, ax=ax)
+            ax = plotting.plot_scatter(xdata, ydata, ax=ax)
+            self._format_plot(ax)
+            figures = [ax.get_figure()]
+        else:
+            figures = None
+
+        return best_fit, figures
+
+    @staticmethod
+    def _fit_quality(fit_freq: float, reduced_chisq: float, fit_freq_err: float):
+        """Method to check the quality of the fit.
+
+        A good fit has:
+            - Have a reduced chi-squared lower than three.
+            - More than a quarter of a full period.
+            - Less than 10 full periods.
```

Comment:
Consider (not urgent, relevant also to other experiments) to soft-code these numbers

Reply:
Perhaps worth opening an issue and aligning everything in one PR.